client/peerset/src/lib.rs

```diff
@@ -168,13 +171,21 @@ impl PeersetHandle {
 	}
 
 	/// Returns the reputation value of the peer.
-	pub async fn peer_reputation(self, peer_id: PeerId) -> Result<i32, ()> {
+	pub fn peer_reputation(&self, peer_id: PeerId) -> impl Future<Output = Result<i32, ()>> {
```

Comment:
Did the `async` actually had to be removed here? (Considering that `async` *is* syntax sugar for `impl Future<...>`.)

Reply:
`async` with `&self` should work?

Reply:
The following (`async` with `&self`) borrows the `self` immutably for the whole lifetime of the produced by the async-method future.

```rust
pub async fn peer_reputation(&self, peer_id: PeerId) -> Result<i32, ()> {
	let (tx, rx) = oneshot::channel();
	let _ = self.tx.unbounded_send(Action::PeerReputation(peer_id, tx));
	// The channel can only be closed if the peerset no longer exists.
	rx.await.map_err(|_| ())
}
```

Which leads to the following:
```
error[E0502]: cannot borrow `*self` as mutable because it is also borrowed as immutable
   --> client/network/src/request_responses.rs:579:22
    |
579 | ...                   let handler = self.new_handler_with_replacement(protocol, handler);
    |                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ mutable borrow occurs here
...
610 | ...                   let get_peer_reputation = self.peerset.peer_reputation(peer);
    |                                                 ---------------------------------- immutable borrow occurs here
...
623 | ...                       get_peer_reputation,
    |                           ------------------- cast requires that `self.peerset` is borrowed for `'static`

error: lifetime may not live long enough
   --> client/network/src/request_responses.rs:623:9
    |
430 |         &mut self,
    |         - let's call the lifetime of this reference `'1`
...
623 |                                 get_peer_reputation,
    |                                 ^^^^^^^^^^^^^^^^^^^ cast requires that `'1` must outlive `'static`
```

which I am not sure how to solve, except for de-sugaring the `peer_reputation` method so, that the produced future clearly does not borrow `self`.

Reply:
Okay ty. The calling code is weird, but that is not part of this pr.